src/esp/assets/managers/AttributesManagerBase.h

```diff
@@ -0,0 +1,334 @@
+
+  /**
+   * @brief Maps string keys to attributes templates
+   */
+  std::map<std::string, std::shared_ptr<T>> templateLibrary_;
+
+  /**
+   * @brief Maps all object attribute IDs to the appropriate handles used
+   by
+   * lib
+   */
+  std::map<int, std::string> templateLibKeyByID_;
+
+ public:
+  ESP_SMART_POINTERS(AttributesManager)
+
+};  // class AttributesManager
+
+/////////////////////////////
+// Class Template Definitions
+// template <typename T>
+// AttributesManager<T>::AttributesManager() {}
+
+// template <typename T>
+// AttributesManager<T>::~AttributesManager() {}
+
+template <typename T>
+std::string AttributesManager<T>::getRandomTemplateHandlePerType(
+    const std::map<int, std::string>& mapOfHandles,
+    const std::string& type) const {
+  int numVals = mapOfHandles.size();
+  if (numVals == 0) {
+    LOG(ERROR) << "Attempting to get a random " << type
+               << "object template handle but none are loaded; Aboring";
+    return "";
+  }
+  int randIDX = rand() % numVals;
+
+  std::string res;
+  for (std::pair<std::map<int, std::string>::const_iterator, int> iter(
+           mapOfHandles.begin(), 0);
+       (iter.first != mapOfHandles.end() && iter.second <= randIDX);
+       ++iter.first, ++iter.second) {
+    res = iter.first->second;
+  }
+  return res;
+}  // getRandomTemplateHandlePerType
+
+template <typename T>
+std::vector<std::string>
+AttributesManager<T>::getTemplateHandlesBySubStringPerType(
+    const std::map<int, std::string>& mapOfHandles,
+    const std::string& subStr,
+    bool contains) const {
+  std::vector<std::string> res;
+  // if empty return empty vector
+  if (mapOfHandles.size() == 0) {
+    return res;
+  }
+  // if search string is empty, return all values
+  if (subStr.length() == 0) {
+    for (auto elem : mapOfHandles) {
+      res.push_back(elem.second);
+    }
+    return res;
+  }
+  // build search criteria
+  std::string strToLookFor(subStr);
+
+  int strSize = strToLookFor.length();
+  // force lowercase
+  std::transform(strToLookFor.begin(), strToLookFor.end(), strToLookFor.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+
+  for (std::map<int, std::string>::const_iterator iter = mapOfHandles.begin();
+       iter != mapOfHandles.end(); ++iter) {
+    std::string key(iter->second);
+    // be sure that key is big enough to search in (otherwise find has undefined
+    // behavior)
+    if (key.length() < strSize) {
+      continue;
+    }
+    // force lowercase - search is case insensitive
+    std::transform(key.begin(), key.end(), key.begin(),
+                   [](unsigned char c) { return std::tolower(c); });
+    bool found = (std::string::npos != key.find(strToLookFor));
+    if (found == contains) {
+      // if found and searching for contains, or not found and searching for not
+      // contains
+      res.push_back(iter->second);
+    }
+  }
+  return res;
+}  // getTemplateHandlesBySubStringPerType
```

Comment:
I remember we were discussing putting the function definitions into the cpp file, since it's quite a huge bulk of code and the set of types is limited and known in advance.

Did that not work, after all? Or what's the reason it's all in the header?

Reply:
These were put here to avoid future linker errors.  Otherwise I would have to put all possible typename (attributes types) declarations at the bottom, and this would then break when more attributes were added to the project.